proof/crefine/AARCH64/IsolatedThreadAction.thy

```diff
@@ -1572,19 +1570,23 @@ lemma thread_actions_isolatableD:
 lemma tcbSchedDequeue_rewrite:
   "monadic_rewrite True True (obj_at' (Not \<circ> tcbQueued) t) (tcbSchedDequeue t) (return ())"
   apply (simp add: tcbSchedDequeue_def)
-   apply (wp_pre, monadic_rewrite_symb_exec_l_known False, simp)
-    apply (rule monadic_rewrite_refl)
-   apply (wpsimp wp: threadGet_const)+
+  apply wp_pre
+  apply monadic_rewrite_symb_exec_l
+    apply (monadic_rewrite_symb_exec_l_known False, simp)
+     apply (rule monadic_rewrite_refl)
+    apply (wpsimp wp: threadGet_const)+
   done
 
+(* FIXME: improve automation here *)
 lemma switchToThread_rewrite:
   "monadic_rewrite True True
        (ct_in_state' (Not \<circ> runnable') and cur_tcb' and obj_at' (Not \<circ> tcbQueued) t)
        (switchToThread t)
        (do Arch.switchToThread t; setCurThread t od)"
   apply (simp add: switchToThread_def Thread_H.switchToThread_def)
   apply (monadic_rewrite_l tcbSchedDequeue_rewrite, simp)
-   apply (rule monadic_rewrite_refl)
+  apply (repeat_unless \<open>rule monadic_rewrite_refl\<close> monadic_rewrite_symb_exec_l)
+      apply wpsimp+
```

Comment:
I'm curious what's happening here that stops the refl being the right way to go. In some cases you do really need to do a bunch of symbolic execution, make a claim that the return values are identical, and then wpsimp your way to showing that. The `repeat_unless` seems a bit odd, but it's probably doing the same as `monadic_rewrite_symb_exec_l+, rule monadic_rewrite_refl`. Does the `repeat_unless` gain anything?

Reply:
I'll open this up tomorrow and check. I'll report back

Reply:
`repeat_unless` is actually doing something for us here, which is to say that `monadic_rewrite_symb_exec_l+, rule monadic_rewrite_refl` is not what we want. 

The LHS of the `monadic_rewrite` has a function with some getters and some `assert/stateAssert`s and then afterwards, the two functions are exactly the same. If we just let `monadic_rewrite_symb_exec_l+` run, then it would consume too much of the LHS and then we don't get a perfect matching for `monadic_rewrite_refl` to then apply. I thought that maybe there would be an easy way to get one of the custom methods to carry this out for us, but I'm not sure which one it should be. Do you have any other guesses? I've tried things with both `monadic_rewrite_l` and `monadic_rewrite_symb_exec_l` and can't get anything working

Reply:
And we do have `monadic_rewrite_single_pass`, which uses `repeat_unless` internally, but that doesn't seem to be intended as a "user level" method for our apply scripts

Reply:
I see now, yeah. A comment would've helped here ("get rid of getters and asserts on LHS" or somesuch). It looks like you want something like a "symbolically execute on left until X, then refl", but without the automation to do that, you're doing "try refl, and if that doesn't work symbolically execute on left, repeat".  I'll take a look at whether I've handled something like this in a nicer way, but for now this is fine. I think "my way" would've been to add a comment like I suggested, then do `repeat 4 monadic_rewrite_symb_exec_l` or something, then refl. With the comment, these feel roughly equivalent.

Reply:
`monadic_rewrite_single_pass` is designed for when one of the sides is schematic, and you want to step over the non-schematic side until a rule applies... I don't think it'll help here, it sounds like your RHS isn't schematic anyway.

Reply:
I think I prefer the `repeat_unless` over the `repeat` with a hard coded number, just in case we add more asserts to something and the number changes. 

Would `(* strip LHS of getters and asserts until LHS and RHS are the same *)` be a good comment? I can go with your comment verbatim if you like, or if you find a nicer way, I can do that

Reply:
Your comment is more useful, happy with that and keeping current version. I can't think of a better way of achieving this, and understand what you mean about more getters/asserts, so let's keep your version as well. If you feel like there's a principle here that automation could handle better, let me know.